substrate/client/network/sync/src/engine.rs

```diff
@@ -770,6 +782,25 @@ where
 			}
 		}
 
+		// Retreive warp sync target block header just before polling `ChainSync`
+		// to make progress as soon as we receive it.
+		if let Some(ref mut target_block_header_rx) = self.warp_sync_target_block_header_rx {
+			match target_block_header_rx.poll_unpin(cx) {
```

Comment:
This has to be reworked somehow, otherwise we won't be able to convert `SyncingEngine` to use `tokio::select!`. Maybe a stream that yields only one event, remembers that it has emitted that event and then keeps returning `Poll::Pending`?

Reply:
I don't know if such stream exists. But anyway, it seems we can replace a pinned `rx` future with something like `Fuse::terminated()` once we get a value out of the channel, can't we? This would be something similar to what's done here, but more `select!`-compatibe.

Reply:
I was thinking we could implement that stream, shouldn't be too hard but what you're suggesting sounds sensible as well and I'd go with it. We probably have to visit this `Option<oneshot::Receiver>` business anyway at some point because it doesn't look too elegant and hopefully there is some other way of implementing this.

Reply:
As was suggested by @arkpar in the original issue https://github.com/paritytech/cumulus/issues/1619#issuecomment-1321769298, we could have added a dedicated method to `NetworkService` to set the target block instead of using the channel. But I don't like that approach either, because it's more cumbersome an error-prone (it's less obvious that one should call some special method compared to having a stream as part of the initialization params).

Reply:
I only now remembered that AFAIR I've had problems with fused futures and `tokio::select!` but could be a phantom memory. Either way, your solutions looks good but if in the future we run into trouble, we may have to revisit.